test/export/StructureDefinitionExporter.test.ts

```diff
@@ -618,6 +618,110 @@ describe('StructureDefinitionExporter', () => {
+        'http://hl7.org/fhir/StructureDefinition/MolecularSequence'
+      )
+    );
+
+    expect(constrainedHasMemberFoo.type).toHaveLength(1);
+    expect(constrainedHasMemberFoo.type[0]).toEqual(
+      new ElementDefinitionType('Reference').withTargetProfiles(
+        'http://example.com/StructureDefinition/Bar',
+        'http://hl7.org/fhir/StructureDefinition/QuestionnaireResponse',
+        'http://hl7.org/fhir/StructureDefinition/MolecularSequence'
+      )
+    );
+
+    expect(constrainedHasMemberBar.type).toHaveLength(1);
+    expect(constrainedHasMemberBar.type[0]).toEqual(
+      new ElementDefinitionType('Reference').withTargetProfiles(
+        'http://example.com/StructureDefinition/Foo',
+        'http://hl7.org/fhir/StructureDefinition/QuestionnaireResponse',
+        'http://hl7.org/fhir/StructureDefinition/MolecularSequence'
+      )
+    );
+  });
+
+  it('should apply correct OnlyRule with circular FSHy parent', () => {
+    const profile1 = new Profile('Foo');
+    profile1.parent = 'Observation';
+    const profile2 = new Profile('Bar');
+    profile2.parent = 'Foo';
+
+    doc.profiles.set(profile2.name, profile2);
+
+    const rule = new OnlyRule('hasMember[Observation]');
+    rule.types = [{ type: 'Bar', isReference: true }];
+    profile1.rules.push(rule);
+
+    exporter.exportStructDef(profile1);
+    const sdFoo = exporter.profileDefs.find(def => def.id === 'Foo');
+    const sdBar = exporter.profileDefs.find(def => def.id === 'Bar');
+    const baseStructDef = resolve('Observation');
+
+    expect(sdFoo.baseDefinition).toBe('http://hl7.org/fhir/StructureDefinition/Observation');
+    expect(sdBar.baseDefinition).toBe('http://example.com/StructureDefinition/Foo');
+
+    const baseHasMember = baseStructDef.findElement('Observation.hasMember');
+    const constrainedHasMemberFoo = sdFoo.findElement('Observation.hasMember');
+    const constrainedHasMemberBar = sdBar.findElement('Observation.hasMember');
+
+    expect(baseHasMember.type).toHaveLength(1);
+    expect(baseHasMember.type[0]).toEqual(
+      new ElementDefinitionType('Reference').withTargetProfiles(
+        'http://hl7.org/fhir/StructureDefinition/Observation',
+        'http://hl7.org/fhir/StructureDefinition/QuestionnaireResponse',
+        'http://hl7.org/fhir/StructureDefinition/MolecularSequence'
+      )
+    );
+
+    expect(constrainedHasMemberFoo.type).toHaveLength(1);
+    expect(constrainedHasMemberFoo.type[0]).toEqual(
+      new ElementDefinitionType('Reference').withTargetProfiles(
+        'http://example.com/StructureDefinition/Bar',
+        'http://hl7.org/fhir/StructureDefinition/QuestionnaireResponse',
+        'http://hl7.org/fhir/StructureDefinition/MolecularSequence'
+      )
+    );
+
+    expect(constrainedHasMemberBar.type).toHaveLength(1);
+    expect(constrainedHasMemberBar.type[0]).toEqual(
+      new ElementDefinitionType('Reference').withTargetProfiles(
+        'http://hl7.org/fhir/StructureDefinition/Observation',
```

Comment:
Uh oh.  I think this might be wrong...  If `Bar` has parent `Foo`, and `Foo` says `* hasMember[Observation] only Reference(Bar)` -- then shouldn't `Bar` inherit that constraint?  Meaning that the above URL should *actually* be `http://example.com/StructureDefinition/Bar`?

Reply:
So I think I know why this happens.  Here's the breakdown:

1. First it exports `Foo`, so it starts processing `Foo`...
2. It gets to `* hasMember[Observation] only Bar` so it tries to resolve `Bar`
3. `Bar` hasn't been exported yet, so it starts processing `Bar`
4. The first thing it does when processing `Bar` is resolve its parent: `Foo`
5. It gets back the _incomplete_ `Foo`, which doesn't yet have the `hasMember` constraint applied
6. Since it uses this incomplete `Foo` as its starting point it misses the `hasMember` constraint

Reply:
So what do we do?  Well...

`Foo` really only needed the metadata of `Bar` -- because it just needed its `url`.  The problem is that even though that is all we needed, we generated the whole `Bar` then and there and stored its results.  If we had only processed `Bar`'s metadata and didn't generate all of `Bar` then and there, we would have been fine.

So... we might be back to adding an argument to `resolve` that says `metadataOnly` -- in which case `resolve` would only process the metadata and not push `Bar` at that point...  Does that make sense?